docker/docker_tools.py

```diff
@@ -0,0 +1,541 @@
+#
+# No loops
+# --------
+# Unlike earlier approaches, this script doesn't cause the container to restart if something goes wrong. Instead, it catches all errors in a try/except block then stops executing so you can see what happened.
+#
+#
+# venvs
+# =====
+# All Python installs are placed in a virtual environment -- ``/root/venv`` and also (for dev builds) in a venv managed by Poetry. Before running Python scripts, be sure to activate the relevant venv.
+#
+#
+# Imports
+# =======
+# These are listed in the order prescribed by PEP 8.
+#
+# Standard library
+# ----------------
+from pathlib import Path
+import re
+import subprocess
+import sys
+from time import sleep
+from traceback import print_exc
+from typing import Dict, Tuple
+from textwrap import dedent
+
+# Local application
+# -----------------
+# The working directory of this script.
+wd = Path(__file__).resolve().parent
+sys.path.append(str(wd / "../tests"))
+from ci_utils import chdir, env, xqt, is_linux
+
+# Third-party
+# -----------
+try:
+    import click
+except ImportError:
+    print("Installing click...")
+    xqt(
+        f"{sys.executable} -m pip install --upgrade pip",
+        f"{sys.executable} -m pip install --upgrade click",
+    )
+    import click
+
+
+# CLI interface
+# =============
+@click.group()
+def cli():
+    pass
+
+
+# ``build`` command
+# =================
+@cli.command()
+# Allow users to pass args directly to the underlying ``docker build`` command -- see the `click docs <https://click.palletsprojects.com/en/8.0.x/arguments/#option-like-arguments>`_.
+@click.argument("passthrough", nargs=-1, type=click.UNPROCESSED)
+@click.option("--arm/--no-arm", default=False, help="Install the ARMv7 toolchain.")
+@click.option("--dev/--no-dev", default=False, help="Install tools needed for development with the Runestone.")
+@click.option("--pic24/--no-pic24", default=False, help="Install tools needed for development with the PIC24/dsPIC33 family of microcontrollers.")
+@click.option("--rust/--no-rust", default=False, help="Install the Rust toolchain.")
+@click.option("--tex/--no-tex", default=False, help="Instal LaTeX and related tools.")
+def build(arm: bool, dev: bool, passthrough: Tuple, pic24: bool, tex: bool, rust: bool):
+    """
+    When executed outside a Docker build, build a Docker container for the Runestone webservers.
+
+        PASSTHROUGH: These arguments are passed directly to the underlying "docker build" command. To pass options to this command, prefix this argument with "--". For example, use "docker_tools.py build -- -no-cache" instead of "docker_tools.py build -no-cache" (which produces an error).
+
+    Inside a Docker build, install all dependencies as root.
+    """
+
+    assert is_linux, "This program is only capable of running on Ubuntu/Debian."
```

Comment:
The only impediment to running on macos is removing this line. 😄

Reply:
Wow, fantastic! Fixed with pleasure.